### ISSUE TITLE
[DEVEL] Nonexistent request revert during unit test 

### ISSUE THREAD

**Issue by rayyan808** [smartcontractkit/chainlink#7370]
**Description**
I am running the `VRFCoordinatorV2MockLocal` in my local environment using hardhat. All seemed to work well until I ported my codebase into the `ERC2535` a.k.a Diamond Proxy standard. Essentially my stack-trace using `hardhat/console.sol` narrows it down to a failure to set the `s_requests[].subId` correctly. But when adding logs to the `VRFCoordinatorV2MockLocal` code itself, I could not figure out where my code is causing issues, seems to be a failure on the mock contract to assign the value to the mapping?

``
**Your Environment**
Ubuntu 20 with yarn and hardhat starter kit integration
**Basic Information**

**Steps to Reproduce**
Deployment Steps
```
log("Obtaining Link Token Contract...")
    linkToken = await ethers.getContract("LToken")
    log("Obtaining VRFCoordinatorV2Mock... ")
    VRFCoordinatorV2Mock = await ethers.getContract("VRFCoordinatorV2MockLocal")

    vrfCoordinatorAddress = VRFCoordinatorV2Mock.address
    linkTokenAddress = linkToken.address
    log("Creating local VRF Subscription..")
    const fundAmount = networkConfig[chainId]["fundAmount"]
    const transaction = await VRFCoordinatorV2Mock.createSubscription()
    const transactionReceipt = await transaction.wait(1)
    subscriptionId = ethers.BigNumber.from(transactionReceipt.events[0].topics[1])
    log("Funding local VRF SUBSCRIPTION ID: %s with AMOUNT: %s..", subscriptionId, fundAmount)
    await VRFCoordinatorV2Mock.fundSubscription(subscriptionId, fundAmount)
log("Adding Diamond Contract to VRFCoordinator Consumers")
  await VRFCoordinatorV2Mock.addConsumer(subscriptionId, elyxrDiamond.address)


  log("Transfering 90000000000 TEST LINK to Chainlink Facet")

  tx = await linkToken.transfer(elyxrDiamond.address, 90000000000) 
  receipt = await tx.wait()
  if (!receipt.status) {
    throw Error(`Link transfer failed: ${tx.hash}`)
  }
```
Add the following logs to `VRFCoordinatorV2MockLocal`
```
function fulfillRandomWords(uint256 _requestId, address _consumer) external {
    uint256 startGas = gasleft();
    console.log("(VRFCoordinatorMock-fulfillRandomWords) Recieved arguments requestId: %s, consumer: %s \n", _requestId, _consumer);
    if (s_requests[_requestId].subId == 0) {
      console.log("(VRFCoordinatorMock-fulfillRandomWords) s_requests[%s].subId = 0 \n", _requestId);
      revert("nonexistent request");
    }
    Request memory req = s_requests[_requestId];

    uint256[] memory words = new uint256[](req.numWords);
    for (uint256 i = 0; i < req.numWords; i++) {
      words[i] = uint256(keccak256(abi.encode(_requestId, i)));
    }

    VRFConsumerBaseV2 v;
    bytes memory callReq = abi.encodeWithSelector(v.rawFulfillRandomWords.selector, _requestId, words);
    (bool success, ) = _consumer.call{gas: req.callbackGasLimit}(callReq);

    uint96 payment = uint96(BASE_FEE + ((startGas - gasleft()) * GAS_PRICE_LINK));
    if (s_subscriptions[req.subId].balance < payment) {
      revert InsufficientBalance();
    }
    s_subscriptions[req.subId].balance -= payment;
    delete (s_requests[_requestId]);
    emit RandomWordsFulfilled(_requestId, _requestId, payment, success);
  }
 function requestRandomWords(
    bytes32 _keyHash,
    uint64 _subId,
    uint16 _minimumRequestConfirmations,
    uint32 _callbackGasLimit,
    uint32 _numWords
  ) external override returns (uint256) {
    console.log("(VRFMockCoordinator-requestRandomWords) Recieved request from address %s with argument subId: %s", msg.sender, _subId);
    if (s_subscriptions[_subId].owner == address(0)) {
      console.log("INVALID SUBSCRIPTION GIVEN TO MOCK COORDINATOR");
      revert InvalidSubscription();
    }

    uint256 requestId = s_nextRequestId++;
    uint256 preSeed = s_nextPreSeed++;

    s_requests[requestId] = Request({subId: _subId, callbackGasLimit: _callbackGasLimit, numWords: _numWords});
    console.log("s_requests[%s].subId = %s", requestId, _subId);
    emit RandomWordsRequested(
      _keyHash,
      requestId,
      preSeed,
      _subId,
      _minimumRequestConfirmations,
      _callbackGasLimit,
      _numWords,
      msg.sender
    );
    return requestId;
  }
```

Using a `Diamond` contract, perform two `delegatecall` from `Diamond` to `Contract B` to `Contract C`, `Contract C` inherits from `VRFConsumerBaseV2` and performs the request (and successfully returns a valid requestId)

```
  Contract call:       <UnrecognizedContract>
  Transaction:         0xe216ccefed45860fd8652adb4a8e2cdf15d300e9c4c938b441708aaf6c8d897f
  From:                0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266
  To:                  0x2279b7a0a67db372996a5fab50d91eaa73d2ebe6
  Value:               0.5 ETH
  Gas used:            180991 of 12450000000
  Block #23:           0xf43ebcce78c732258686cdf3fc9d5c97d84059299059fae99d68dbe75c9f70ce

  console.log:
    (Diamond Contract) call to facet 0x0dcd1bf9a1b36ce34237eeafef220932846bcd82:
    (Contract B) Calling facet at 0xa51c1fc2f0d1a1b8494ed1fe312d7c3a78ed91c0: 
    (Contract C [VRFConsumerBaseV2 inheritor]) Requesting using sub id: 1
    (VRFMockCoordinator-requestRandomWords) Recieved request from address 0x2279b7a0a67db372996a5fab50d91eaa73d2ebe6 with argument subId: 1
    s_requests[1].subId = 1
    (Diamond Contract) Recieved final Request Id: 1 from Chainlink successfully

eth_chainId
eth_getTransactionByHash
eth_chainId
eth_getTransactionReceipt

```
The key observation here is the output: `s_requests[1].subId = 1`. We now follow steps to perform fulfillment of random words, and compare the values of `s_requests[1].subId`

Since it's a local node, we need to invoke the fulfillment function ourselves. So I use the following code to read an event emission from the transaction, reading the `requestId` and `roller (original msg.sender wallet)` and invoking the function manually:
```
async function performFulfillment2(txReceipt, coordinator, elyxrDiamondAddress) {
    const resolvedReceipt = await txReceipt.wait()
    //console.log(resolvedReceipt.events[1].args)
    const requestId = resolvedReceipt.events[1].args[0]//.requestId 
    const eventInvokerAddress = resolvedReceipt.events[1].args[1]//.roller
    console.log("Mimicking VRFCoordinator callback with requestID(%s) FOR roller: %s.. using Consumer %s", requestId, eventInvokerAddress, elyxrDiamondAddress)
    await coordinator.fulfillRandomWords(requestId, elyxrDiamondAddress)
  
    var obj = { eventAddress: eventInvokerAddress, request: requestId}
    return obj;
  }
```

The logs that are output from the fulfillment function via `yarn hardhat test` show that the `consumer` and `requestId` are correctly emitted in the event.
```
Mimicking VRFCoordinator callback with requestID(1) FOR address: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266.. using Consumer 0x2279b7a0a67db372996a5fab50d91eaa73d2ebe6
```

We then see the `hardhat node` output as follows:
```
eth_chainId
eth_estimateGas
  Contract call:       <UnrecognizedContract>
  From:                0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266
  To:                  0xc6e7df5e7b4f2a278906862b61205850344d4e7d
  Value:               0 ETH

  console.log:
    (VRFCoordinatorMock-fulfillRandomWords) Recieved arguments requestId: 1, consumer: 0x2279b7a0a67db372996a5fab50d91eaa73d2ebe6
    s_requests[1].subId == 0

  Error: VM Exception while processing transaction: reverted with reason string 'nonexistent request'
```

as you can see, now the `s_requests[1].subId` is returned as `0`, executing the reversion code:
```
if (s_requests[_requestId].subId == 0) {
      console.log("(VRFCoordinatorMock-fulfillRandomWords) s_requests[%s].subId = 0 \n", _requestId);
      revert("nonexistent request");
    }
```
**Additional Information**
I also attempted to perform the initialization on-chain by invoking the following function:
```
   function linkToChainlink() public onlyOwner {
        AppStorage storage s = LibDiamond.appStorage();
        s.s_subscriptionId = s.coordinatorInterface.createSubscription();
        s.coordinatorInterface.addConsumer(s.s_subscriptionId, address(this));
        console.log("Added Diamond as consumer subID: %s address: %s", s.s_subscriptionId, address(this));
    }
```
I have also tried making `Contract C` the consumer directly, this does not affect the issue.
Any tips or advice is highly appreciated.

**Comment by rgottleber** [comment]
Thank you for sharing so many details. I think I've lost the thread of the actual issue you are facing. Could you clarify what the issue is you see and if it's an issue with Chainlink itself? 

Thanks

**Comment by rayyan808** [comment]
> Thank you for sharing so many details. I think I've lost the thread of the actual issue you are facing. Could you clarify what the issue is you see and if it's an issue with Chainlink itself?
> 
> Thanks

`requestRandomWords` runs correctly in the `VRFCoordinatorV2Mock`, and returns `requestId = 1`. When `fulfillRandomWords` is invoked, the transaction reverts with `nonexistent request`. The issue I see is that the `requestRandomWords` will assign `s_requests[1].subId = 1`. But when `fulfillRandomWords` is called, the revert occurs because `s_requests[1].subId = 0`. 
I also perform `await txReceipt.wait()` when `requestRandomWords` occurs, in order to make sure the transaction has been mined. This is definitely the case (transaction is mined) as I read an event from the same transaction to obtain the requestId and the 'roller' address. I hope this clarifies your confusion.
To summarize: `revert('nonexistent request)` is called when it should not be. I don't know if this is a chainlink issue, but I also don't see how it is an issue on my side hopefully this thread gives me more clarity on the issue

**Comment by rgottleber** [comment]
Thank you for the clarification and for opening the issue. We will take a look and get back to you on this issue.

**Comment by rayyan808** [comment]
Any updates?

**Comment by rayyan808** [comment]
As it turns out, I was using `ethers.getContractAt` for all contracts, but using `ethers.getContract` for the Coordinator. This caused a mismatch in the address.
